### PR TITLE
Reduce object allocation/boxing

### PR DESCRIPTION
**Pull request by Basewq**
# PR Details

These commits are aimed at reducing memory allocations, which in turn reduces the garbage collector being called to clean up the objects.

## Description

* Reduce the number of structs that get boxed due to missing IEquatable inheritance when used as a `Dictionary` key.
* Avoid structs being boxed (ie. another object allocation) in the object array in `ParameterCollection` when the currently stored value is the same as the incoming value.
* ~Reduce the number of times `RenderMesh` is created by fixing the mesh count logic check.~
* Use concrete classes when calling enumerator to get the struct enumerable to avoid object allocation.

## Additional Notes
For the `ParameterCollection` fix, I am unsure whether the second check is required or if it could just go into the same logic as the equality test.
To play it safe, if `<T>` is an object type (ie. a proper object reference), it'll always be assigned because there might be some edge case where if the object is internally equal but *not* reference equals, then a programmer manipulates the newly assigned object after it is set, then the programmer expects it to be changed.
However, this does pay a cost of having the second check.

~You are free to reject the `RenderMesh` commit as it currently behaves on "broken" logic (and essentially trades one quirk for another).
Currently, it is broken for models that have materials that do multiple passes (eg. opaque pass + transparent pass) and it'll always go through the RegenerateMeshes code path because mesh count logic always fails.
This change ensures it counts the meshes correctly, however if a user somehow replaces the meshes but retains the same count, it will not render with the new mesh.
Perhaps this is ok, and we need an explicit force regenerate flag or processor method?~

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: sources/engine/Xenko.Engine/Updater/UpdateEngine.cs [hunk 1]
```diff
@@ -738,6 +738,27 @@ public override string ToString()
...
+
+            public bool Equals(UpdateKey other)
+            {
+                return Owner == other.Owner && Name == other.Name;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (ReferenceEquals(null, obj)) return false;
+                return obj is UpdateKey && Equals((UpdateKey)obj);
```
Comment (Eideren) [inline sources/engine/Xenko.Engine/Updater/UpdateEngine.cs]: Can be simplified to ``return obj is UpdateKey key && Equals(key);`` the null test above isn't necessary as the is test already returns false for nulls.

File: sources/engine/Xenko.Rendering/Extensions/IndexExtensions.cs [hunk 2]
```diff
@@ -309,6 +309,17 @@ public EdgeKeyAEN(EdgeAEN edge)
...
 
+            public bool Equals(EdgeKeyAEN other)
+            {
+                return PositionIndex0 == other.PositionIndex0 && PositionIndex1 == other.PositionIndex1;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (ReferenceEquals(null, obj)) return false;
+                return obj is EdgeKeyAEN && Equals((EdgeKeyAEN)obj);
```
Comment (Eideren) [inline sources/engine/Xenko.Rendering/Extensions/IndexExtensions.cs]: Same here.

File: sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs [hunk 3]
```diff
@@ -144,7 +144,7 @@ private void CheckMeshes(ModelComponent modelComponent, RenderModel renderModel)
                 if (model != null)
                 {
                     // Number of meshes changed in the model?
-                    if (model.Meshes.Count != renderModel.Meshes.Length)
+                    if (model.Meshes.Count != renderModel.UniqueMeshCount)
```
Comment (Eideren) [inline sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: From what I understood of the issue you mentioned on the post this would be resolved by comparing the two collections but you would rather not pay the perf cost of doing so ? Perhaps this could be resolved by introducing a collection with a ``version`` variable, you would test against the ``version`` variable instead of ``UniqueMeshCount`` and, if we really want to be thorough, do a collection comparison afterwards ?
If you think this would resolve the issue, @xen2 could we add a ``version`` to FastList or would you rather have another collection type for this purpose ?

File: sources/engine/Xenko.Engine/Updater/UpdateEngine.cs (hunk 1 above)
Comment (Basewq) [reply sources/engine/Xenko.Engine/Updater/UpdateEngine.cs]: Ok, I copied it from an existing code. I assume this means we can use C# 7 features (I think I saw some in use already). Will make this change.

File: sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs (hunk 3 above)
Comment (Basewq) [reply sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: Yes, the current faulty logic is trying to detect changes in the Meshes list, I think.
If you're willing for a proper breaking change, I think changing Meshes from a `List` to a `TrackingCollection` on the `Model` and having some sort of `RequiresMeshRegen` bool will suffice.
Even if they end up adding the exact same elements, I don't think it matters too much if it regenerates the mesh (especially since they're probably going through the regen code currently and it shouldn't happen every update frame anyway).
However, I think this will need some sort of serialization upgrade code or something? All I know is Model is one of the most complicated asset types and I do not dare to mess with that, haha.

File: sources/engine/Xenko.Engine/Updater/UpdateEngine.cs (hunk 1 above)
Comment (Basewq) [reply sources/engine/Xenko.Engine/Updater/UpdateEngine.cs]: Please advise if you'd prefer me to amend the existing commit and force push instead of making another commit.

File: sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs (hunk 3 above)
Comment (Eideren) [reply sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: Let's see what @xen2 has to say

Comment (xen2) [reply sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: I think even if we were to compare collection, it wouldn't be enough as `Mesh` and `Model` is not immutable (which would help a lot for some specific classes such as `Mesh` to simplify a lot of those clunky checks, however it would require non-trivial changes to runtime/editor).

For example, `Mesh.MaterialIndex` could still change afterwise, as well as `Model.Materials`.
That's why the second step is not checking not the `Mesh` (either as a class reference or all its content), but only the important part requiring a rebuild (actual material with all the indirections resolved).

Comment (Basewq) [reply sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: Ok, perhaps this change is not safe enough to make at this point in time.
I am ok with this not included in the pull.
Do you want me to remove this commit from this pull, or leave every thing as is and you'll sort it out from the rest of the commits?

Comment (Eideren) [reply sources/engine/Xenko.Engine/Rendering/ModelRenderProcessor.cs]: Please remove it, I'll merge afterwards if xen doesn't do it before I have time to :)

**Comment by Basewq** [comment]
Ok, removed the mesh regen commit, squashed Eideren & Kryptos-FR's advices into their respective commits and rebased.

**Comment by Eideren** [comment]
Thanks !
